### PR TITLE
[#9] Support ring.core.protocols/StreamableResponseBody

### PR DESCRIPTION
**Pull request by lispyclouds**
This does work but not sure if this is the best way? Addresses #9

### CODE REVIEW

**Comment by mpenet** [comment]
Thanks, that seems like a good addition. I just wish it didn't bring all these dependencies with it, but I don't think we have a choice.

**Comment by lispyclouds** [comment]
Yeah thought about that but I guess this is the only way to be drop-in compatible with existing Ring code?

**Comment by mpenet** [comment]
I suppose it could be made a sort of a "provided" dependency and inject the extra code only if ring is detected on the classpath.

**Comment by lispyclouds** [comment]
Is it a fair assumption that if the user has reified that protocol, they have ring-core in their dependencies? It's certainly true for me. If yes, we can drop it.

**Comment by mpenet** [comment]
You could do like aleph does: https://github.com/clj-commons/aleph/blob/e64c2c541b6080a739d58187164f9c11345c95b5/src/aleph/http/client_middleware.clj#L32

Check if ring is loadable, if it is we extend the BodyWriter proto for Object with the write-body-to-stream, otherwise we extend id as it is in master now.

**Comment by lispyclouds** [comment]
Added the conditional loading.

**Comment by lispyclouds** [comment]
This approach causes linter warnings about namespace `p` not being found, but the works fine. The alternative like aleph does is:
```clojure
(write-body! [o server-response]
  (if ring-core-loaded
    (let [StreamableResponseBody @(ns-resolve 'ring.core.protocols 'StreamableResponseBody)
          write-body-to-stream (ns-resolve 'ring.core.protocols 'write-body-to-stream)]
      (when (satisfies? StreamableResponseBody o)
        (write-body-to-stream o nil (.outputStream server-response))))
    (.send ^ServerResponse server-response o)))
```
is it worth it?

**Comment by mpenet** [comment]
I think that should be done as a separate extend-protocol call that only happens if ring is present, so that's no overhead without. I can take over this one if you want (I am working on websocket support at the moment, so it will have to wait a bit).

**Comment by lispyclouds** [comment]
I'll give it a go.

**Comment by lispyclouds** [comment]
Is that what you had in mind?

**Comment by mpenet** [comment]
Almost! I think the extend form should be either build dynamically or evaled to work with aot, otherwise upon aot if ring is not present it would blow up.

**Comment by lispyclouds** [comment]
I was under the impression that compilation should work due to the when check which should be false when ring-core is absent? Also I thought the body of function implementing the protocol would be dynamically evaled?

**Comment by mpenet** [comment]
Nevermind I didn't see you called ns-resolve. That should work. Could you add a tests show it supports it now?

**Comment by lispyclouds** [comment]
added tests

File: src/s_exp/mina/response.clj [hunk 1]
```diff
@@ -7,6 +7,12 @@
            (io.helidon.webserver.http ServerResponse)
            (java.io FileInputStream InputStream OutputStream)))
 
+(def ^:no-doc ring-core-loaded
```
Comment (mpenet) [inline src/s_exp/mina/response.clj:10]: small nit: rename to `ring-core-loaded?`

Comment (lispyclouds) [reply src/s_exp/mina/response.clj]: done, i tend to have that for pred functions.

**Comment by mpenet** [comment]
Thanks!

**Pull request merged**